File: Decyx/Decyx.py
# Python extension that leverages Anthropic's Claude to assist in reverse engineering and binary analysis.
# @author Phil Sajdak
# @category AI Analysis
# @keybinding Shift R
# @menupath
# @toolbar

from ghidra.framework.preferences import Preferences
from decyx.config import CLAUDE_MODELS, SKIP_PROMPT_CONFIRMATION
from decyx.api import get_response_from_claude
from decyx.decompiler import decompile_function, decompile_callers
from decyx.utils import (
    apply_selected_suggestions, apply_line_comments, apply_explanation,
    prepare_prompt
)
from decyx.gui import *

def get_api_key(preferences):
    """
    Retrieve the API key from Ghidra's preferences. If the key does not exist,
    prompt the user to input the Anthropic Claude API key and store it in the preferences.
    """
    api_key = preferences.getProperty("ANTHROPIC_API_KEY")
    if not api_key:
        api_key = askString("API Key", "Enter your Anthropic Claude API key:", "")
        if api_key:
            preferences.setProperty("ANTHROPIC_API_KEY", api_key)
            preferences.store()
            print "Anthropic API Key stored in {}.".format(preferences.getFilename())
    return api_key

def get_callers_code(func, current_program, monitor):
    """
    Get the decompiled code of the functions that call the current function.
    This is useful for additional context in the analysis.
    """
    callers = func.getCallingFunctions(monitor)
    if not callers:
        return None
    
    print "Found {} caller(s) for the current function.".format(len(callers))
    selected_callers = show_caller_selection_dialog(list(callers), current_program, monitor)
    return decompile_callers(selected_callers, current_program, monitor) if selected_callers else None

def process_action(action, func, current_program, monitor, api_key, model, callers_code):
    """
    Process a specific action on the decompiled function, sending the data to the Claude API and applying the response.
    Actions can include renaming, retyping variables, adding explanations, and inserting line comments.
    """
    decompiled_code, variables = decompile_function(func, current_program, monitor, annotate_addresses=(action == 'line_comments'))
    if not decompiled_code or not variables:
        print "Failed to obtain decompiled code or variable information for {}.".format(action)
        return False

    prompt = prepare_prompt(decompiled_code, variables, action=action, callers_code=callers_code)
    final_prompt = prompt if SKIP_PROMPT_CONFIRMATION else show_prompt_review_dialog(prompt, "Review and Edit Prompt ({})".format(action.replace('_', ' ').title()))
    if not final_prompt:
        print "Prompt review cancelled by user."
        return False

    is_explanation = action == 'explanation'
    response = get_response_from_claude(final_prompt, api_key, model, monitor, is_explanation=is_explanation)
    if not response:
        print "Failed to get {} from Claude API.".format(action.replace('_', ' '))
        return False

    if action == 'rename_retype':
        selected_suggestions = show_suggestion_dialog(response, variables, state.getTool())
        if selected_suggestions:
            apply_selected_suggestions(func, response, selected_suggestions, state.getTool())
        else:
            print "Operation cancelled by user after receiving suggestions."
            return False
    elif action == 'explanation':
        apply_explanation(func, response)
    elif action == 'line_comments':
        apply_line_comments(func, response)

    return True

def main():
    """
    The main entry point of the script. Responsible for gathering API keys, 
    selecting models and actions, and processing the actions on the current function.
    """
    api_key = get_api_key(Preferences)
    if not api_key:
        print "API key is required to proceed."
        return

    if len(CLAUDE_MODELS) == 1:
        model = CLAUDE_MODELS[0]
        print "Using the only available model: {}".format(model)
    else:
        model = show_model_select_dialog(CLAUDE_MODELS)
        if not model:
            print "Model selection cancelled by user."
            return

    selected_actions = show_action_select_dialog()
    if not selected_actions:
        print "No actions selected. Exiting."
        return

    func = getFunctionContaining(currentAddress)
    if not func or not currentProgram or not monitor:
        print "Required context is missing."
        return

    callers_code = get_callers_code(func, currentProgram, monitor)
    print "Callers' code {}.".format("included for additional context" if callers_code else "not included")

    for action in selected_actions:
        print "Processing action: {}".format(action)
        if not process_action(action, func, currentProgram, monitor, api_key, model, callers_code):
            print "Failed to process action: {}".format(action)
            return

    print "Decyx operations completed successfully."

if __name__ == "__main__":
    main()

File: Decyx/Decyx.java
// Java extension that leverages Anthropic's Claude to assist in reverse engineering and binary analysis.
// @author Phil Sajdak
// @category AI Analysis
// @keybinding Shift R
// @menupath
// @toolbar

import com.google.gson.JsonObject;
import decyx.ClaudeApi;
import decyx.DecyxConfig;
import decyx.DecompilerHelper;
import decyx.DecompilerHelper.DecompileResult;
import decyx.SuggestionUtils;
import decyx.dialogs.ActionSelectionDialog;
import decyx.dialogs.CallerSelectionDialog;
import decyx.dialogs.ModelSelectionDialog;
import decyx.dialogs.PromptReviewDialog;
import decyx.dialogs.SuggestionDialog;
import ghidra.app.script.GhidraScript;
import ghidra.framework.preferences.Preferences;
import ghidra.program.model.listing.Function;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class Decyx extends GhidraScript {

    @Override
    protected void run() throws Exception {
        // Get API key
        String apiKey = getApiKey();
        if (apiKey == null || apiKey.isEmpty()) {
            println("API key is required to proceed.");
            return;
        }

        // Select model
        String model;
        List<String> models = DecyxConfig.CLAUDE_MODELS;
        if (models.size() == 1) {
            model = models.get(0);
            println("Using the only available model: " + model);
        } else {
            model = ModelSelectionDialog.showDialog(models);
            if (model == null) {
                println("Model selection cancelled by user.");
                return;
            }
        }

        // Select actions
        List<String> selectedActions = ActionSelectionDialog.showDialog();
        if (selectedActions.isEmpty()) {
            println("No actions selected. Exiting.");
            return;
        }

        // Get current function context
        Function func = getFunctionContaining(currentAddress);
        if (func == null || currentProgram == null || monitor == null) {
            println("Required context is missing.");
            return;
        }

        // Get callers' code for additional context
        Map<String, String> callersCode = getCallersCode(func);
        println("Callers' code " + (callersCode != null ? "included for additional context" : "not included") + ".");

        // Process each selected action
        for (String action : selectedActions) {
            println("Processing action: " + action);
            if (!processAction(action, func, apiKey, model, callersCode)) {
                println("Failed to process action: " + action);
                return;
            }
        }

        println("Decyx operations completed successfully.");
    }

    /**
     * Retrieves the API key from Ghidra's preferences, prompting the user if not stored.
     */
    private String getApiKey() throws Exception {
        String apiKey = Preferences.getProperty("ANTHROPIC_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            apiKey = askString("API Key", "Enter your Anthropic Claude API key:", "");
            if (apiKey != null && !apiKey.isEmpty()) {
                Preferences.setProperty("ANTHROPIC_API_KEY", apiKey);
                Preferences.store();
                println("Anthropic API Key stored in " + Preferences.getFilename() + ".");
            }
        }
        return apiKey;
    }

    /**
     * Gets the decompiled code of functions that call the current function.
     */
    private Map<String, String> getCallersCode(Function func) {
        Set<Function> callers = func.getCallingFunctions(monitor);
        if (callers == null || callers.isEmpty()) {
            return null;
        }

        println("Found " + callers.size() + " caller(s) for the current function.");
        List<Function> callerList = new ArrayList<>(callers);
        List<Function> selectedCallers = CallerSelectionDialog.showDialog(callerList, currentProgram, monitor);

        if (selectedCallers == null || selectedCallers.isEmpty()) {
            return null;
        }

        return DecompilerHelper.decompileCallers(selectedCallers, currentProgram, monitor);
    }

    /**
     * Processes a specific action on the decompiled function.
     */
    private boolean processAction(String action, Function func, String apiKey, String model,
                                   Map<String, String> callersCode) {
        boolean annotateAddresses = "line_comments".equals(action);
        DecompileResult decompResult = DecompilerHelper.decompileFunction(func, currentProgram, monitor, annotateAddresses);
        if (decompResult == null || decompResult.code == null || decompResult.variables == null) {
            println("Failed to obtain decompiled code or variable information for " + action + ".");
            return false;
        }

        String prompt = SuggestionUtils.preparePrompt(decompResult.code, decompResult.variables, action, callersCode);
        String finalPrompt;
        if (DecyxConfig.SKIP_PROMPT_CONFIRMATION) {
            finalPrompt = prompt;
        } else {
            String title = "Review and Edit Prompt (" + action.replace("_", " ").substring(0, 1).toUpperCase() +
                action.replace("_", " ").substring(1) + ")";
            finalPrompt = PromptReviewDialog.showDialog(prompt, title);
        }

        if (finalPrompt == null) {
            println("Prompt review cancelled by user.");
            return false;
        }

        boolean isExplanation = "explanation".equals(action);
        Object response = ClaudeApi.getResponseFromClaude(finalPrompt, apiKey, model, monitor, isExplanation);
        if (response == null) {
            println("Failed to get " + action.replace("_", " ") + " from Claude API.");
            return false;
        }

        switch (action) {
            case "rename_retype":
                JsonObject suggestions = (JsonObject) response;
                JsonObject selectedSuggestions = SuggestionDialog.showDialog(suggestions, decompResult.variables, state.getTool());
                if (selectedSuggestions != null) {
                    SuggestionUtils.applySelectedSuggestions(func, suggestions, selectedSuggestions, state.getTool());
                } else {
                    println("Operation cancelled by user after receiving suggestions.");
                    return false;
                }
                break;

            case "explanation":
                SuggestionUtils.applyExplanation(func, (String) response);
                break;

            case "line_comments":
                SuggestionUtils.applyLineComments(func, (JsonObject) response);
                break;
        }

        return true;
    }
}
